### PR TITLE
Added the ability to update CSM model state in jigsaw

### PR DESCRIPTION
**Pull request by jessemapel**
## Description

Adds the ability to write updated CSM model states to cubes in jigsaw.

## Related Issue

#4410 

## Motivation and Context

This allows us to store the updated model state from jigsaw for use in future applications or for export to other SETs such as Socet GXP.

## How Has This Been Tested?

Tested locally on CTX test data set. The updated BLOB is on the cube with the jigged comment. Inspected USGSCSM state string JSON and it contained the updated parameters. Pre and post runs of campt on images also showed updates consistent with the results of the jigsaw solution.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [ ] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/control/objs/BundleAdjust/BundleAdjust.cpp [hunk 1]
```diff
@@ -2845,6 +2844,27 @@ namespace Isis {
...
+   */
+  QString BundleAdjust::modelState(int i) {
+    Camera *imageCam = m_controlNet->Camera(i);
+    if (imageCam->GetCameraType() != Camera::Csm) {
+      QString msg = "Cannot get model state for image [" + toString(i) +
+                    "] because it is not a CSM camera model.";
+      throw IException(IException::Programmer, msg, _FILEINFO_);
+    }
+
+    CSMCamera *csmCamera = dynamic_cast<CSMCamera*>(imageCam);
```
Comment (jessemapel) [inline isis/src/control/objs/BundleAdjust/BundleAdjust.cpp:2863]: I am not happy about having the dynamic_cast in BundleAdjust and not using the Observation, but attempts to use the polymorphism of the observation resulting in strange issues with the indices in the control network not matching the indices in the observations. I also ran into issues with who has the correct camera model and simultaneous open/close of the cube. This approach side-steps all of those issues at the cost of the cast happening here.

**Review by krlberry — APPROVED** [review]
This all looks good to me!

File: isis/tests/FunctionalTestsJigsaw.cpp [hunk 2]
```diff
@@ -1594,7 +1594,7 @@ TEST_F(MiniRFNetwork, FunctionalTestJigsawRadar) {
 }
 
 
-TEST_F(CSMNetwork, FunctionalTestCSMNetwork) {
+TEST_F(CSMNetwork, unctionalTestJigsawCSM) {
```
Comment (krlberry) [inline isis/tests/FunctionalTestsJigsaw.cpp]: Could you put the 'F' back in?

**Review by krlberry — CHANGES_REQUESTED** [review]
One character typo! Then I'll go back to approving the PR.

**Review by krlberry — APPROVED** [review]

**Pull request merged**